pkg/controller/master-controller-manager/kubeone/controller.go

```diff
@@ -231,91 +229,220 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
+	if err := r.initiateImportAction(ctx, log, externalCluster); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	clusterStatus := externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status
+	externalClusterProvider, err := kubernetesprovider.NewExternalClusterProvider(r.ImpersonationClient, r.Client)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	manifestSecret := &corev1.Secret{}
+	if err := r.Get(ctx, types.NamespacedName{Namespace: externalCluster.Spec.CloudSpec.KubeOne.ManifestReference.Namespace, Name: externalCluster.Spec.CloudSpec.KubeOne.ManifestReference.Name}, manifestSecret); err != nil {
+		return reconcile.Result{}, fmt.Errorf("can not retrieve kubeone manifest secret: %w", err)
+	}
+	currentManifest := manifestSecret.Data[resources.KubeOneManifest]
+
+	_, err = r.upgradeAction(ctx, log, externalCluster, externalClusterProvider, currentManifest, clusterStatus)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err = r.checkPodStatusIfExists(ctx, log, externalCluster, UpgradeControlPlaneAction); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	_, err = r.migrateAction(ctx, log, externalCluster, externalClusterProvider, currentManifest, clusterStatus)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err = r.checkPodStatusIfExists(ctx, log, externalCluster, MigrateControlPlaneAction); err != nil {
+		return reconcile.Result{}, err
+	}
 
 	return reconcile.Result{}, nil
 }
 
-func (r *reconciler) checkPodStatus(ctx context.Context, log *zap.SugaredLogger, pod *corev1.Pod, externalCluster *kubermaticv1.ExternalCluster) error {
+func returnWantedContainerRuntime(currentManifest []byte) (string, error) {
+	cluster := &kubeonev1beta2.KubeOneCluster{}
+	if err := yaml.UnmarshalStrict(currentManifest, cluster); err != nil {
+		return "", fmt.Errorf("failed to decode manifest secret data: %w", err)
+	}
+	if cluster.ContainerRuntime.Docker != nil {
+		return resources.ContainerRuntimeDocker, nil
+	} else if cluster.ContainerRuntime.Containerd != nil {
+		return resources.ContainerRuntimeContainerd, nil
+	}
+	return "", nil
+}
+
+func (r *reconciler) checkPodStatus(ctx context.Context,
+	log *zap.SugaredLogger,
+	pod *corev1.Pod,
+	externalCluster *kubermaticv1.ExternalCluster,
+	action string,
+) error {
 	log.Debugw("Checking kubeone pod status", "Pod", pod.Name)
-	for pod.Status.Phase == corev1.PodSucceeded {
+	if pod.Status.Phase == corev1.PodSucceeded {
 		oldexternalCluster := externalCluster.DeepCopy()
 		// update kubeone externalcluster status.
 		externalCluster.Spec.CloudSpec.KubeOne.ClusterStatus.Status = kubermaticv1.StatusRunning
 
 		if err := r.Patch(ctx, externalCluster, ctrlruntimeclient.MergeFrom(oldexternalCluster)); err != nil {
-			return fmt.Errorf("failed to add kubeconfig reference to %s: %w", externalCluster.Name, err)
+			return fmt.Errorf("failed to update cluster status %s: %w", externalCluster.Name, err)
```

Comment:
All of the errors returned by this function should not contain the cluster name. Instead, if you need, either attach it to the logger when you log the error (further up the stacktrace) or add the cluster name at a higher level in the stacktrace.

Meaning don't do

```go
func foobar(name string) error {
   if bla {
      return fmt.Errorf("%s is bad, because: foo", name)
   }
 
   if foobar {
      return fmt.Errorf("%s is bad, because schwupsdiwups", name)
   }

   return nil
}

if err := foobar("xrstf"); err != nil {
    return err
}
```

but instead do

```go
func foobar(name string) error {
   if bla {
      return errors.new("foo")
   }
 
   if foobar {
      return errors.New("schwupsdiwups")
   }

   return nil
}

if err := foobar("xrstf"); err != nil {
    return fmt.Errorf("xrstf is invalid: %w", err)
}
```